File: ios/Packages/FatwaBotKit/Sources/ConfigKit/ConfigService.swift
import CoreKit
import Foundation
import NetworkingKit

/// Config sync per docs/features/config-sync.md. Offline-first: `current` is
/// always readable (cache → empty snapshot whose consumers fall back to bundled
/// values); `refresh` updates layers independently and persists validated results.
public actor ConfigService {
    private let store: ConfigStoring
    private let client: APIClientProtocol
    private let now: @Sendable () -> Date
    private(set) public var current: ConfigSnapshot

    public init(
        store: ConfigStoring,
        client: APIClientProtocol,
        now: @escaping @Sendable () -> Date = { Date() }
    ) {
        self.store = store
        self.client = client
        self.now = now
        self.current = store.load() ?? ConfigSnapshot()
    }

    /// Fetches all layers concurrently. Per-layer failures are silent (spec):
    /// the failed layer keeps its cached value. Returns the layers that changed.
    @discardableResult
    public func refresh(locales: [String]) async -> Set<ConfigLayer> {
        async let configResult = try? client.get(Endpoints.config)
        async let themeResult = try? client.get(Endpoints.theme)
        async let layoutResult = try? client.get(Endpoints.home)

        var packs: [String: StringPack] = [:]
        for locale in locales {
            let since = current.stringPacks[locale]?.version
            if let pack = try? await client.get(Endpoints.strings(locale: locale, sinceVersion: since)) {
                packs[locale] = pack
            }
        }

        var changed: Set<ConfigLayer> = []
        var next = current

        if let config = await configResult, config != next.appConfig {
            next.appConfig = config
            changed.insert(.appConfig)
        }
        if let theme = await themeResult, theme != next.theme {
            next.theme = theme
            changed.insert(.theme)
        }
        if let layout = await layoutResult, layout != next.homeLayout {
            next.homeLayout = layout
            changed.insert(.homeLayout)
        }
        for (locale, pack) in packs where pack != next.stringPacks[locale] {
            next.stringPacks[locale] = pack
            changed.insert(.strings)
        }

        if let country = countryCode(of: next), country != "*" || next.prayerDefaults == nil {
            if let defaults = try? await client.get(Endpoints.prayerDefaults(country: country)),
               defaults != next.prayerDefaults {
                next.prayerDefaults = defaults
                changed.insert(.prayerDefaults)
            }
        }

        if !changed.isEmpty || next.fetchedAt == nil {
            next.fetchedAt = now()
            current = next
            store.save(next)
        }
        return changed
    }

    /// Server pack overlay → nil (caller falls back to bundled Localizable, then key).
    public func string(_ key: String, locale: String) -> String? {
        current.stringPacks[locale]?.strings[key]
    }

    /// Flag gating: unknown flag = false; disabled = false; min_app_version gate
    /// honored (percentage rollout lands with identity bucketing in M3).
    public func isEnabled(_ flag: String, appVersion: String) -> Bool {
        guard let entry = current.appConfig?.flags[flag], entry.enabled else { return false }
        if let minimum = entry.rollout?.minAppVersion {
            return SemVer.isVersion(appVersion, atLeast: minimum)
        }
        return true
    }

    public func staleness() -> TimeInterval? {
        current.fetchedAt.map { now().timeIntervalSince($0) }
    }

    private func countryCode(of snapshot: ConfigSnapshot) -> String? {
        // M1: country comes from device region; injected later with location work.
        Locale.current.region?.identifier ?? "*"
    }
}

File: android/core/config/src/main/kotlin/com/fatwabot/core/config/ConfigService.kt
package com.fatwabot.core.config

import com.fatwabot.core.common.AppConfig
import com.fatwabot.core.common.HomeLayout
import com.fatwabot.core.common.PrayerDefaults
import com.fatwabot.core.common.SemVer
import com.fatwabot.core.common.ServerTheme
import com.fatwabot.core.common.StringPack
import com.fatwabot.core.network.ApiClientProtocol
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json

/**
 * Config sync (ADR-0011 client half) — mirror of iOS ConfigKit.ConfigService.
 * Offline-first: [current] is always readable (cache → empty snapshot whose
 * consumers fall back to bundled values); [refresh] updates layers independently.
 */
class ConfigService(
    private val store: ConfigStoring,
    private val client: ApiClientProtocol,
    private val nowEpochSeconds: () -> Long,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val mutex = Mutex()

    @Volatile
    var current: ConfigSnapshot = store.load() ?: ConfigSnapshot()
        private set

    /** Fetches all layers concurrently; per-layer failures are silent (spec). */
    suspend fun refresh(locales: List<String>): Set<ConfigLayer> = mutex.withLock {
        coroutineScope {
            val configJob = async { fetch("v1/config") { json.decodeFromString<AppConfig>(it) } }
            val themeJob = async { fetch("v1/config/theme") { json.decodeFromString<ServerTheme>(it) } }
            val layoutJob = async { fetch("v1/home") { json.decodeFromString<HomeLayout>(it) } }

            val packs = locales.mapNotNull { locale ->
                val since = current.stringPacks[locale]?.version
                val query = if (since != null) mapOf("since_version" to since.toString()) else emptyMap()
                fetch("v1/config/strings/$locale", query) {
                    // Server returns {"up_to_date": true} when nothing newer.
                    if (it.contains("\"up_to_date\"")) null else json.decodeFromString<StringPack>(it)
                }?.let { locale to it }
            }.toMap()

            val changed = mutableSetOf<ConfigLayer>()
            var next = current

            configJob.await()?.takeIf { it != next.appConfig }?.let {
                next = next.copy(appConfig = it); changed += ConfigLayer.APP_CONFIG
            }
            themeJob.await()?.takeIf { it != next.theme }?.let {
                next = next.copy(theme = it); changed += ConfigLayer.THEME
            }
            layoutJob.await()?.takeIf { it != next.homeLayout }?.let {
                next = next.copy(homeLayout = it); changed += ConfigLayer.HOME_LAYOUT
            }
            packs.forEach { (locale, pack) ->
                if (pack != next.stringPacks[locale]) {
                    next = next.copy(stringPacks = next.stringPacks + (locale to pack))
                    changed += ConfigLayer.STRINGS
                }
            }

            if (changed.isNotEmpty() || next.fetchedAtEpochSeconds == null) {
                next = next.copy(fetchedAtEpochSeconds = nowEpochSeconds())
                current = next
                store.save(next)
            }
            changed
        }
    }

    /** Server pack overlay → null (caller falls back to bundled strings, then key). */
    fun string(key: String, locale: String): String? =
        current.stringPacks[locale]?.strings?.get(key)

    /** Unknown flag = false; disabled = false; min_app_version gate honored. */
    fun isEnabled(flag: String, appVersion: String): Boolean {
        val entry = current.appConfig?.flags?.get(flag) ?: return false
        if (!entry.enabled) return false
        val min = entry.rollout?.minAppVersion ?: return true
        return SemVer.isVersionAtLeast(appVersion, min)
    }

    fun homeLayout(): HomeLayout? = current.homeLayout
    fun prayerDefaults(): PrayerDefaults? = current.prayerDefaults

    /** GET [path]; any failure (network/parse) → null (silent per spec). */
    private suspend inline fun <T> fetch(
        path: String,
        query: Map<String, String> = emptyMap(),
        decode: (String) -> T?,
    ): T? = runCatching { decode(client.getRaw(path, query)) }.getOrNull()
}
